### PR TITLE
Datastream unavailable exception metadata

### PR DESCRIPTION
**Pull request by albertzaharovits**
The goal of this PR is to reuse the "resource unavailable" exception conditions,
between when wildcards are expanded and when they're not.
It's a (big) step in that direction, rather than a complete work.
I have stopped short of a complete implementation because
I became aware of a required behavior change that I think needs
 analyzed and discussed in advance, so that I can confidently proceed.

I'll explain commit-by-commit:

* https://github.com/elastic/elasticsearch/pull/91461/commits/19a41282ded243935fd8d532611f9e7ab91e0f60 When wildcards are expanded, explicitly naming datastreams when the context disallows datastreams, and also when `ignore_unavailable` is `false`, will generate a "not found" exception, see: https://github.com/elastic/elasticsearch/blob/6e60e6ac9ee27e7203ee65a504c023eb5333611a/server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java#L1274 
However, in the same context that doesn't allow datastreams, but without expanding wildcards, datastreams are always ignored, even if `ignore_unavailable` is `false`. This commit fixes that.

* https://github.com/elastic/elasticsearch/pull/91461/commits/938d2c03429fb721ddb9fa6df3d9995e563b951b **Is the behavior change that I think needs to be discussed**. Some "index not found" exceptions have a metadata field `es.excluded_ds: true`. The field is only used in a single place, and only for update requests (which don't support datastreams, and only target a single index): https://github.com/elastic/elasticsearch/blob/d5b45846127456487dfdf59c2ffe13e90356e468/server/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java#L146 . It is however serialized, so it is visible to callers.
   * Curently, the metadata field is attached when an expression evalutes to the **empty concrete indices list**, and at least a datastreams was ignored in the process (either because `ignore_unavailable` is `true` or because, as described above (and fixed), datastreams are always ignored when wildcards are not expanded).
   * The proposed PR will attach the metadata field whenever a datastream is explicitly named in a context that disallows datastreams and `ignore_unavailable` is `false`.
   * Both approaches are equivalent with respect to the internal usage of the field (used only in the context of the update request, that disallows datastreams, refers to a single shard, and `ignore_unavailable` is `false`), but the metadata field is visible to callers.
   * I want to introduce this change is because I want to separate the layers that 1) resolve an expression to a list of index, alias, and datastream names (which will be reused in Security), and 2) expand the elements of that list to concrete indices (which will remain in Core). I cannot achieve the separation if the layer that expands to concrete indices (2) has to know about ignored datastreams, which are handled in the other layer.
* https://github.com/elastic/elasticsearch/pull/91461/commits/d1f11d43bf21797d27b7da7cd2883a08d792f719 Reuses the method that generates "not found" exceptions. In this context I realised that for a **single** "unavailable" name, sometimes the exception was constructed with `infe.setResources("index_or_alias", expression);` and sometimes with `infe.setResources("index_expression", expression);`; I opted to always use `index_or_alias`.
* https://github.com/elastic/elasticsearch/pull/91461/commits/6b2dc8e83700f0ffcfaaca87323ef932d7a220e3 This shows the direction I'm going to, I will iterate on it to encapsulate the handling of unavailable resources inside the `resolveExpressions` method.

------------

**Is the slight behavior change (circumstances when the "not found" exceptions contain a metadata field, without impacting any other behaviors, but that can be observed by callers in the serialized exception, as detailed above) a concern for anybody?**

### CODE REVIEW

**Comment by albertzaharovits** [comment]
@elasticmachine run elasticsearch-ci/part-1

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java [hunk 1]
```diff
@@ -346,27 +346,18 @@ Index[] concreteIndices(Context context, String... indexExpressions) {
...
-                    assert options.expandWildcardExpressions() == false;
-                    throw aliasesNotSupportedException(expression);
-                } else {
-                    continue;
-                }
+                continue;
             } else if (indexAbstraction.isDataStreamRelated() && context.includeDataStreams() == false) {
-                excludedDataStreams = true;
                 continue;
             }
```
Comment (albertzaharovits) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java:363]: All of this is going to be moved, in a follow-up PR, inside the `resolveExpression` method on line 339.

**Comment by grcevski** [comment]
This sounds good to me. I can't find anyone referring to `es.excluded_ds` and it's not documented, I think we can safely say it was only for the internal use of our own transport action. Do you know if it's mentioned in any client specs? Again, I can't find it, but I just want to make sure I didn't miss a place.

**Comment by albertzaharovits** [comment]
>  Do you know if it's mentioned in any client specs? Again, I can't find it, but I just want to make sure I didn't miss a place.

No, I can't find references to it (it eventually gets serialized in the exception body of the response as `excluded_ds`). The rest specs that I'm aware of (`rest-api-spec` project) are not concerned with responses at all. In addition I know we should change the "shape" of responses, and I believe we're laxer with error response bodies (though I admit I'm not sure of specific guidelines). I have asked the lang clients team for advice.

**Comment by albertzaharovits** [comment]
> I have asked the lang clients team for advice.

@swallez confirmed on another channel that client code doesn't look at exception metadata fields.

**Review by grcevski — APPROVED** [review]
LGTM!

**Review by jakelandis — APPROVED** [review]
LGTM

Can you add the bug label to the PR ? I think at least the first commit fixes a bug and the second one is arguable a bug too.

**Comment by albertzaharovits** [comment]
@elasticsearchmachine update branch

**Pull request merged**